test/functional/ex_cmds/echo_spec.lua

```diff
@@ -11,31 +11,48 @@ local dedent = helpers.dedent
 local command = helpers.command
 local exc_exec = helpers.exc_exec
 local redir_exec = helpers.redir_exec
+local pcall_err = helpers.pcall_err
+local matches = helpers.matches
 
-describe(':echo', function()
+local function StringErr(str, quote)
+  local ret = pcall_err(command, 'echoerr '..(quote and "'"..str:gsub("'", "''").."'" or str))
+  return ret:gsub('^Vim%(echoerr%):', '', 1)
+end
+
+describe(':echo :echomsg :echoerr', function()
   before_each(function()
     clear()
     source([[
       function String(s)
         return execute('echo a:s')[1:]
       endfunction
+      function StringMsg(s)
+        return execute('echomsg a:s')[1:]
+      endfunction
     ]])
   end)
 
   describe('used to represent floating-point values', function()
     it('dumps NaN values', function()
-      eq('str2float(\'nan\')', eval('String(str2float(\'nan\'))'))
+      eq("str2float('nan')", eval("String(str2float('nan'))"))
+      eq("str2float('nan')", eval("StringMsg(str2float('nan'))"))
+      eq("str2float('nan')", StringErr("str2float('nan')"))
```

Comment:
in all of the new test cases, it looks like the inputs are the same, and the expected result is always the same. 

So instead of adding cases for echo/echomsg/echoerr, we could have one function which asserts equivalence before returning the result. Something like:

```lua
-- Test stringify behavior of :echo, :echomsg, :echoerr.
local function StringX(str, quote)
  local ret = pcall_err(command, 'echoerr '..(quote and "'"..str:gsub("'", "''").."'" or str))
  -- Assert same behavior for :echo.
  eq(ret, eval("execute('echo "..str.."')[1:]")
  -- Assert same behavior for :echomsg.
  eq(ret, eval("execute('echomsg "..str.."')[1:]")
  return ret:gsub('^Vim%(echoerr%):', '', 1)
end
```

Reply:
Those asserts won't always work because of lambdas and that `:echo` and `echomsg`  output are not always the same. `f_string()` is non-recursive for self-referencing dicts, lists.

Reply:
wouldn't it still save 50+ lines of code?